### DISCUSSION TITLE
Stable Marriage

### DISCUSSION THREAD

**Discussion by hunsicke** [moj-analytical-services/splink#1602] [Q&A]
It seems like after doing the linkage, we will want an implementation of the stable marriage algorithm to identify best linkage pairs.  Has that been done?  Is there a reason that is a bad way to do things?  If I implement this in pyspark, would it be worthwhile to integrate into Splink capabilities?

**Comment by ADBond** [comment]
We are certainly interested in [having a method for one-to-one matching](https://github.com/moj-analytical-services/splink/issues/251). So far we just have this [ad-hoc piece of SQL](https://moj-analytical-services.github.io/splink/topic_guides/splink_fundamentals/querying_splink_results.html#__code_0) which implements a basic solution which has some deficiencies.

I think the potential difficulty with the stable marriage approach is the kind of case outlined in [this comment](https://github.com/moj-analytical-services/splink/issues/251#issuecomment-986843984) - if I understand correctly the stable matching would be 'scenario 1', which one would probably consider to be suboptimal when considering the probabilities instead of simply the order of preference (i.e. we probably prefer the unstable 'scenario 2').

NB: in cases where we don't need one-to-one matching then we use [`linker.cluster_pairwise_predictions_at_threshold`](https://moj-analytical-services.github.io/splink/linkerpred.html#splink.linker.Linker.cluster_pairwise_predictions_at_threshold) to run the 'connected components' algorithm to identify matching record sets.

**Comment by RobinL** [comment]
Thanks very much for looking into this - I had guessed it was problem a well known problem, but had never done any research so hadn't come across the 'stable marriage' algo.

I think it would be excellent to have this available on the PySpark side because it's a very general user need.

My initial thoughts:
1. At the moment, all core algorithms can be run against all backends (Spark, Duckdb, Athena, etc.) because they're all written in SQL
2. If the stable marriage algorithm has a (reasonably efficient) SQL implementation it could definitely be something we could want to include in Splink
3. If it's not something that's straightforward to implement in SQL (and would e.g. be much simpler to phrase using alternative PySpark functionality), then maybe a separate splink extension package may be more appropriate.  that is, a separate pip-installable package that contains the PySpark implementation, probablly/possibly assuming data in the format of splink.predict() outputs.  We would certainly be happy to signpost the existence of this from the Splink docs.

Thoughts?  If you think the pyspark implementation would be relatively simple (relatively few lines of code), we could also consider integrating a pyspark-only solution into Splink.  I would just be loathe to add (and therefore support) lots of pyspark specific code, which would be inconsistent with the rest of the codebase.  I hope that makes sense
